bindings/java/src/test/java/org/openucx/jucx/UcpListenerTest.java

```diff
@@ -35,10 +49,98 @@ public void testCreateUcpListener() {
             ipv6Listener.close();
         } catch (UcxException ex) {
 
-        } finally {
-            worker.close();
-            context.close();
         }
 
+        worker.close();
+        context.close();
+    }
+
+    static Stream<NetworkInterface> getInterfaces() {
+        try {
+            return Collections.list(NetworkInterface.getNetworkInterfaces()).stream()
+                .filter(iface -> {
+                    try {
+                        return iface.isUp() && !iface.isLoopback();
+                    } catch (SocketException e) {
+                        return false;
+                    }
+                });
+        } catch (SocketException e) {
+            return Stream.empty();
+        }
+    }
+
+    /**
+     * Iterates over network interfaces and tries to bind and create listener
+     * on a specific socket address.
+     */
+    static UcpListener tryBindListener(UcpWorker worker, UcpListenerParams params) {
+        UcpListener result = null;
+        List<InetAddress> addresses = getInterfaces().flatMap(iface ->
+            Collections.list(iface.getInetAddresses()).stream())
+            .collect(Collectors.toList());
+        for (InetAddress address : addresses) {
+            try {
+                result = worker.newListener(
+                    params.setSockAddr(new InetSocketAddress(address, port)));
+                break;
+            } catch (UcxException ex) {
+
+            }
+        }
+        assertNotNull("Could not find socket address to start UcpListener", result);
+        return result;
+    }
+
+    @Test
+    public void testConnectionHandler() {
+        UcpContext context1 = new UcpContext(new UcpParams().requestRmaFeature()
+            .requestTagFeature());
+        UcpContext context2 = new UcpContext(new UcpParams().requestRmaFeature()
+            .requestTagFeature());
+        UcpWorker worker1 = context1.newWorker(new UcpWorkerParams());
+        UcpWorker worker2 = context2.newWorker(new UcpWorkerParams());
+
+        AtomicLong connRequest = new AtomicLong(0);
+
+        // Create listener and set connection handler
+        UcpListenerParams listenerParams = new UcpListenerParams()
+            .setConnectionHandler(connectionRequest -> connRequest.set(connectionRequest));
+        UcpListener listener = tryBindListener(worker1, listenerParams);
+
+        UcpEndpoint clientToServer = worker2.newEndpoint(new UcpEndpointParams()
+            .setSocketAddress(listener.getAddress()));
+
+        while (connRequest.get() == 0) {
+            worker1.progress();
+            worker2.progress();
+        }
+
+        UcpEndpoint serverToClient = worker1.newEndpoint(
```

Comment:
maybe add another workerN to test connection request migration from worker1 (where listener is created) to workerN?